src/snappy_device_agents/devices/maas2/maas_storage.py

```diff
@@ -0,0 +1,546 @@
+        alloc_pct_values = collections.defaultdict(int)
+
+        for dev in self.device_list:
+            if dev["type"] == "partition":
+                # add pct together (default to 0 if unnused)
+                alloc_pct_values[dev["parent_disk"]] += dev.get("alloc_pct", 0)
+
+        for dev_id, alloc_pct in alloc_pct_values.items():
+            if alloc_pct > 100:
+                raise MaasStorageError(
+                    "The total percentage of the partitions on disk "
+                    f"'{dev_id}' exceeds 100."
+                )
+
+    def create_partition_sizes(self):
+        """Calculate actual partition size to write to disk."""
+        self._validate_alloc_pct_values()
+
+        for dev in self.device_list:
+            if dev["type"] == "partition":
+                # find corresponding block device
+                for block_dev in self.node_info:
+                    if block_dev["id"] == self.block_ids[dev["parent_disk"]]:
+                        # get the total size of the block device in bytes
+                        total_size = int(block_dev["size"])
+                        break
+
+                if "alloc_pct" in dev:
+                    # avoid under-allocating space
+                    dev["size"] = str(
+                        math.ceil((total_size * dev.get("alloc_pct", 0)) / 100)
+                    )
+                else:
+                    if "size" not in dev:
+                        raise ValueError(
+                            f"Partition '{dev['id']}' does not have an "
+                            "alloc_pct or size value."
+                        )
+                    else:
+                        # default to minimum required partition size
+                        dev["size"] = self.convert_size_to_bytes(dev["size"])
+
+    def group_by_type(self):
+        """Group storage devices by type for processing.
+
+        :return: dict with device types as keys and lists of devices as values
+        """
+        devs_by_type = collections.defaultdict(list)
+
+        for dev in self.device_list:
+            devs_by_type[dev["type"]].append(dev)
+
+        return devs_by_type
+
+    def process_by_type(self, devs_by_type):
+        """Process each storage type together in sequence.
+
+        :param devs_by_type: dict with device types as keys and
+            lists of devices as values
+        :raises MaasStorageError: if an error occurs during device processing
+        """
+        # order in which storage types are processed
+        type_order = ["disk", "partition", "format", "mount"]
+        # maps the device type to the method that processes it
+        type_to_method = {
+            "disk": self.process_disk,
+            "partition": self.process_partition,
+            "mount": self.process_mount,
+            "format": self.process_format,
+        }
+        partn_data = {}
+
+        for type_ in type_order:
```

Comment:
device_type

Reply:
Thanks for catching that, from some prototyping and neglected to correct that during cleanup.